File: sensecanvas/client/src/lib/types/widgets.ts
/**
 * SenseCanvas Widget Configuration Types and Schemas
 * Zod schemas and TypeScript types for widget configurations (synced with server models)
 */

import { z } from 'zod';

// Base types and enums
export const WidgetTypeSchema = z.enum(['gauge', 'graph', 'text', 'multi-sensor']);
export const SensorTypeSchema = z.enum(['cpu', 'gpu', 'memory', 'storage', 'network']);
export const ThemeTypeSchema = z.enum(['default', 'cyberpunk', 'gaming', 'minimal', 'rgb']);

export type WidgetType = z.infer<typeof WidgetTypeSchema>;
export type SensorType = z.infer<typeof SensorTypeSchema>;
export type ThemeType = z.infer<typeof ThemeTypeSchema>;

// Position and size schemas
export const PositionSchema = z.object({
  x: z.number().min(0).max(10000),
  y: z.number().min(0).max(10000)
});

export const SizeSchema = z.object({
  width: z.number().min(100).max(800),
  height: z.number().min(100).max(600)
});

export type Position = z.infer<typeof PositionSchema>;
export type Size = z.infer<typeof SizeSchema>;

// Color validation
export const ColorSchema = z.string().regex(/^#[0-9A-Fa-f]{6}$/, 'Invalid hex color');

// Widget style schema
export const WidgetStyleSchema = z.object({
  theme: ThemeTypeSchema,
  colors: z.array(ColorSchema).min(1).max(10),
  opacity: z.number().min(0).max(1),
  borderRadius: z.number().min(0).max(50),
  backgroundImage: z.string().url().optional(),
  fontSize: z.number().min(8).max(48).optional(),
  fontFamily: z.string().optional(),
  borderWidth: z.number().min(0).max(10).optional(),
  borderColor: ColorSchema.optional(),
  shadowEnabled: z.boolean().optional(),
  shadowColor: ColorSchema.optional(),
  shadowBlur: z.number().min(0).max(20).optional(),
  gradientEnabled: z.boolean().optional(),
  gradientDirection: z.enum(['horizontal', 'vertical', 'radial']).optional()
});

export type WidgetStyle = z.infer<typeof WidgetStyleSchema>;

// Alert configuration schemas
export const AlertThresholdsSchema = z.object({
  warning: z.number().min(0).max(100),
  critical: z.number().min(0).max(100)
}).refine(data => data.critical > data.warning, {
  message: "Critical threshold must be higher than warning threshold"
});

export const WidgetAlertsSchema = z.object({
  enabled: z.boolean(),
  thresholds: AlertThresholdsSchema,
  showNotifications: z.boolean().optional(),
  playSound: z.boolean().optional(),
  flashWidget: z.boolean().optional()
});

export type AlertThresholds = z.infer<typeof AlertThresholdsSchema>;
export type WidgetAlerts = z.infer<typeof WidgetAlertsSchema>;

// Type-specific configuration schemas
export const GaugeConfigSchema = z.object({
  minValue: z.number(),
  maxValue: z.number(),
  startAngle: z.number().min(-360).max(360),
  endAngle: z.number().min(-360).max(360),
  arcWidth: z.number().min(1).max(50),
  showValue: z.boolean(),
  showLabel: z.boolean(),
  showTicks: z.boolean(),
  tickInterval: z.number().min(1).max(100),
  unit: z.string().max(10)
}).refine(data => data.maxValue > data.minValue, {
  message: "Max value must be greater than min value"
});

export const GraphConfigSchema = z.object({
  timeRange: z.number().min(1).max(1440), // 1 minute to 24 hours
  maxDataPoints: z.number().min(10).max(1000),
  showGrid: z.boolean(),
  showAxes: z.boolean(),
  lineWidth: z.number().min(1).max(10),
  fillArea: z.boolean(),
  smoothing: z.boolean(),
  yAxisMin: z.number().optional(),
  yAxisMax: z.number().optional()
}).refine(data => {
  if (data.yAxisMin !== undefined && data.yAxisMax !== undefined) {
    return data.yAxisMax > data.yAxisMin;
  }
  return true;
}, {
  message: "Y-axis max must be greater than min"
});

export const TextConfigSchema = z.object({
  format: z.string().min(1).max(100),
  fontSize: z.number().min(8).max(72),
  fontWeight: z.enum(['normal', 'bold', 'light']),
  alignment: z.enum(['left', 'center', 'right']),
  showIcon: z.boolean(),
  iconPosition: z.enum(['left', 'right', 'top', 'bottom']),
  iconSize: z.number().min(8).max(64)
});

export const MultiSensorConfigSchema = z.object({
  sensors: z.array(SensorTypeSchema).min(1).max(10),
  layout: z.enum(['grid', 'list', 'radial']),
  showLabels: z.boolean(),
  showValues: z.boolean(),
  compactMode: z.boolean()
});

export type GaugeConfig = z.infer<typeof GaugeConfigSchema>;
export type GraphConfig = z.infer<typeof GraphConfigSchema>;
export type TextConfig = z.infer<typeof TextConfigSchema>;
export type MultiSensorConfig = z.infer<typeof MultiSensorConfigSchema>;

// Main widget configuration schema
export const WidgetConfigSchema = z.object({
  id: z.string().min(1).max(100),
  type: WidgetTypeSchema,
  title: z.string().min(1).max(50),
  sensorType: SensorTypeSchema,
  position: PositionSchema,
  size: SizeSchema,
  style: WidgetStyleSchema,
  alerts: WidgetAlertsSchema,
  
  // Type-specific configurations (conditional)
  gaugeConfig: GaugeConfigSchema.optional(),
  graphConfig: GraphConfigSchema.optional(),
  textConfig: TextConfigSchema.optional(),
  multiSensorConfig: MultiSensorConfigSchema.optional(),
  
  // Metadata
  createdAt: z.number().min(0),
  updatedAt: z.number().min(0),
  version: z.string().regex(/^\d+\.\d+\.\d+$/, 'Invalid version format'),
  tags: z.array(z.string().max(20)).max(10),
  description: z.string().max(200).optional(),
  author: z.string().max(50).optional(),
  
  // Runtime state (optional, not serialized)
  isSelected: z.boolean().optional(),
  isResizing: z.boolean().optional(),
  isDragging: z.boolean().optional(),
  zIndex: z.number().min(0).max(1000).optional()
}).refine(data => {
  // Ensure type-specific config is present based on widget type
  switch (data.type) {
    case 'gauge':
      return data.gaugeConfig !== undefined;
    case 'graph':
      return data.graphConfig !== undefined;
    case 'text':
      return data.textConfig !== undefined;
    case 'multi-sensor':
      return data.multiSensorConfig !== undefined;
    default:
      return true;
  }
}, {
  message: "Type-specific configuration is required for the selected widget type"
});

export type WidgetConfig = z.infer<typeof WidgetConfigSchema>;

// Grid configuration schema
export const GridSizeSchema = z.object({
  cols: z.number().min(1).max(20),
  rows: z.number().min(1).max(20),
  cellSize: z.number().min(50).max(200)
});

export type GridSize = z.infer<typeof GridSizeSchema>;

// Dashboard layout schema
export const DashboardLayoutSchema = z.object({
  id: z.string().min(1).max(100),
  name: z.string().min(1).max(50),
  description: z.string().max(200),
  widgets: z.array(WidgetConfigSchema).max(50),
  gridSize: GridSizeSchema,
  theme: ThemeTypeSchema,
  backgroundImage: z.string().url().optional(),
  backgroundColor: ColorSchema.optional(),
  
  // Metadata
  createdAt: z.number().min(0),
  updatedAt: z.number().min(0),
  version: z.string().regex(/^\d+\.\d+\.\d+$/, 'Invalid version format'),
  author: z.string().max(50).optional(),
  tags: z.array(z.string().max(20)).max(10),
  
  // Export/Import
  isPublic: z.boolean(),
  shareKey: z.string().optional()
}).refine(data => {
  // Ensure all widget IDs are unique
  const widgetIds = data.widgets.map(w => w.id);
  return widgetIds.length === new Set(widgetIds).size;
}, {
  message: "Widget IDs must be unique within a layout"
});

export type DashboardLayout = z.infer<typeof DashboardLayoutSchema>;

// Widget library item schema
export const WidgetLibraryItemSchema = z.object({
  id: z.string().min(1).max(100),
  name: z.string().min(1).max(50),
  description: z.string().min(1).max(200),
  category: z.string().min(1).max(30),
  tags: z.array(z.string().max(20)).max(10),
  thumbnail: z.string().url().or(z.string().startsWith('data:image/')),
  config: WidgetConfigSchema,
  popularity: z.number().min(0).max(100),
  author: z.string().max(50),
  createdAt: z.number().min(0),
  downloads: z.number().min(0)
});

export type WidgetLibraryItem = z.infer<typeof WidgetLibraryItemSchema>;

// AI generation schemas
export const AiGenerationPromptSchema = z.object({
  id: z.string().min(1).max(100),
  prompt: z.string().min(10).max(500),
  context: z.record(z.any()),
  generatedConfig: z.record(z.any()),
  rating: z.number().min(1).max(5).optional(),
  createdAt: z.number().min(0)
});

export type AiGenerationPrompt = z.infer<typeof AiGenerationPromptSchema>;

// Export/Import schemas
export const WidgetExportSchema = z.object({
  version: z.string().regex(/^\d+\.\d+\.\d+$/, 'Invalid version format'),
  exportedAt: z.number().min(0),
  exportedBy: z.string().max(50).optional(),
  widgets: z.array(WidgetConfigSchema).min(1).max(50),
  layout: DashboardLayoutSchema.optional()
});

export const WidgetImportSchema = z.object({
  data: WidgetExportSchema,
  options: z.object({
    preserveIds: z.boolean().optional(),
    replaceExisting: z.boolean().optional(),
    validateCompatibility: z.boolean().optional()
  }).optional()
});

export type WidgetExport = z.infer<typeof WidgetExportSchema>;
export type WidgetImport = z.infer<typeof WidgetImportSchema>;

// Form validation schemas (for partial updates)
export const WidgetFormDataSchema = WidgetConfigSchema.omit({
  id: true,
  createdAt: true,
  updatedAt: true
});

export const WidgetUpdateSchema = WidgetConfigSchema.partial();

export type WidgetFormData = z.infer<typeof WidgetFormDataSchema>;
export type WidgetUpdate = z.infer<typeof WidgetUpdateSchema>;

// Validation helper functions
export function validateWidgetConfig(config: unknown) {
  const result = WidgetConfigSchema.safeParse(config);
  if (!result.success) {
    return {
      valid: false,
      errors: result.error.issues.map(issue => ({
        field: issue.path.join('.'),
        message: issue.message,
        code: issue.code
      }))
    };
  }
  return { valid: true, data: result.data };
}

export function validateDashboardLayout(layout: unknown) {
  const result = DashboardLayoutSchema.safeParse(layout);
  if (!result.success) {
    return {
      valid: false,
      errors: result.error.issues.map(issue => ({
        field: issue.path.join('.'),
        message: issue.message,
        code: issue.code
      }))
    };
  }
  return { valid: true, data: result.data };
}

export function validateWidgetImport(importData: unknown) {
  const result = WidgetImportSchema.safeParse(importData);
  if (!result.success) {
    return {
      valid: false,
      errors: result.error.issues.map(issue => ({
        field: issue.path.join('.'),
        message: issue.message,
        code: issue.code
      }))
    };
  }
  return { valid: true, data: result.data };
}

// Default configurations
export const DEFAULT_WIDGET_SIZE: Size = { width: 200, height: 200 };
export const MIN_WIDGET_SIZE: Size = { width: 100, height: 100 };
export const MAX_WIDGET_SIZE: Size = { width: 800, height: 600 };

export const DEFAULT_WIDGET_STYLE: WidgetStyle = {
  theme: 'default',
  colors: ['#00ff88', '#ff6b6b'],
  opacity: 1,
  borderRadius: 8,
  fontSize: 14,
  fontFamily: 'Inter, sans-serif',
  borderWidth: 1,
  borderColor: '#e5e7eb',
  shadowEnabled: false,
  shadowColor: '#000000',
  shadowBlur: 4,
  gradientEnabled: false,
  gradientDirection: 'horizontal'
};

export const DEFAULT_ALERTS: WidgetAlerts = {
  enabled: false,
  thresholds: {
    warning: 75,
    critical: 90
  },
  showNotifications: true,
  playSound: false,
  flashWidget: true
};

export const DEFAULT_GAUGE_CONFIG: GaugeConfig = {
  minValue: 0,
  maxValue: 100,
  startAngle: -135,
  endAngle: 135,
  arcWidth: 10,
  showValue: true,
  showLabel: true,
  showTicks: true,
  tickInterval: 20,
  unit: '%'
};

export const DEFAULT_GRAPH_CONFIG: GraphConfig = {
  timeRange: 60, // 1 hour
  maxDataPoints: 100,
  showGrid: true,
  showAxes: true,
  lineWidth: 2,
  fillArea: false,
  smoothing: true
};

export const DEFAULT_TEXT_CONFIG: TextConfig = {
  format: '{value}{unit}',
  fontSize: 24,
  fontWeight: 'normal',
  alignment: 'center',
  showIcon: true,
  iconPosition: 'left',
  iconSize: 24
};

export const DEFAULT_MULTI_SENSOR_CONFIG: MultiSensorConfig = {
  sensors: ['cpu'],
  layout: 'grid',
  showLabels: true,
  showValues: true,
  compactMode: false
};

// Widget type constants
export const WIDGET_TYPES: WidgetType[] = ['gauge', 'graph', 'text', 'multi-sensor'];
export const SENSOR_TYPES: SensorType[] = ['cpu', 'gpu', 'memory', 'storage', 'network'];
export const THEME_TYPES: ThemeType[] = ['default', 'cyberpunk', 'gaming', 'minimal', 'rgb'];

// Utility types
export type WidgetConfiguratorTab = 'library' | 'ai-generate' | 'create-custom';
export type WidgetValidationError = {
  field: string;
  message: string;
  severity: 'error' | 'warning';
};

export interface WidgetConfiguratorState {
  activeTab: WidgetConfiguratorTab;
  currentConfig: Partial<WidgetConfig>;
  validationErrors: WidgetValidationError[];
  isValid: boolean;
  previewData: any;
  libraryItems: WidgetLibraryItem[];
  recentPrompts: AiGenerationPrompt[];
}

// Utility types for forms and API
export type WidgetCreate = Omit<WidgetConfig, 'id' | 'createdAt' | 'updatedAt'>;
export type WidgetPreset = Omit<WidgetConfig, 'id' | 'position' | 'createdAt' | 'updatedAt' | 'isSelected' | 'isResizing' | 'isDragging' | 'zIndex'>;

File: sensecanvas/server/src/models/widgets.py
"""
SenseCanvas Widget Configuration Models
Pydantic models for widget configurations and dashboard layouts.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional, Union
from enum import Enum

from pydantic import BaseModel, Field, validator


class WidgetType(str, Enum):
    """Supported widget types."""
    GAUGE = "gauge"
    GRAPH = "graph"
    TEXT = "text"
    MULTI_SENSOR = "multi-sensor"


class SensorType(str, Enum):
    """Supported sensor types for hardware monitoring."""
    CPU = "cpu"
    GPU = "gpu"
    MEMORY = "memory"
    STORAGE = "storage"
    NETWORK = "network"


class ThemeType(str, Enum):
    """Supported theme types."""
    DEFAULT = "default"
    CYBERPUNK = "cyberpunk"
    GAMING = "gaming"
    MINIMAL = "minimal"
    RGB = "rgb"


class Position(BaseModel):
    """Widget position coordinates."""
    x: float = Field(..., ge=0, le=10000, description="X coordinate in pixels")
    y: float = Field(..., ge=0, le=10000, description="Y coordinate in pixels")


class Size(BaseModel):
    """Widget size dimensions."""
    width: float = Field(..., ge=100, le=800, description="Widget width in pixels")
    height: float = Field(..., ge=100, le=600, description="Widget height in pixels")


class WidgetStyle(BaseModel):
    """Widget styling configuration."""
    theme: ThemeType = Field(..., description="Widget theme")
    colors: List[str] = Field(..., min_items=1, max_items=10, description="Color palette (hex colors)")
    opacity: float = Field(..., ge=0, le=1, description="Widget opacity")
    borderRadius: float = Field(..., ge=0, le=50, description="Border radius in pixels")
    backgroundImage: Optional[str] = Field(None, description="Background image URL")
    fontSize: Optional[float] = Field(None, ge=8, le=48, description="Font size in pixels")
    fontFamily: Optional[str] = Field(None, description="Font family")
    borderWidth: Optional[float] = Field(None, ge=0, le=10, description="Border width in pixels")
    borderColor: Optional[str] = Field(None, regex="^#[0-9A-Fa-f]{6}$", description="Border color")
    shadowEnabled: Optional[bool] = Field(None, description="Enable drop shadow")
    shadowColor: Optional[str] = Field(None, regex="^#[0-9A-Fa-f]{6}$", description="Shadow color")
    shadowBlur: Optional[float] = Field(None, ge=0, le=20, description="Shadow blur radius")
    gradientEnabled: Optional[bool] = Field(None, description="Enable gradient background")
    gradientDirection: Optional[str] = Field(None, regex="^(horizontal|vertical|radial)$", description="Gradient direction")
    
    @validator('colors')
    def validate_colors(cls, colors):
        """Validate color hex codes."""
        import re
        hex_pattern = re.compile(r'^#[0-9A-Fa-f]{6}$')
        for color in colors:
            if not hex_pattern.match(color):
                raise ValueError(f"Invalid hex color: {color}")
        return colors


class AlertThresholds(BaseModel):
    """Alert threshold configuration."""
    warning: float = Field(..., ge=0, le=100, description="Warning threshold percentage")
    critical: float = Field(..., ge=0, le=100, description="Critical threshold percentage")
    
    @validator('critical')
    def validate_critical_higher(cls, critical, values):
        """Ensure critical threshold is higher than warning."""
        if 'warning' in values and critical <= values['warning']:
            raise ValueError("Critical threshold must be higher than warning threshold")
        return critical


class WidgetAlerts(BaseModel):
    """Widget alert configuration."""
    enabled: bool = Field(..., description="Enable alerts for this widget")
    thresholds: AlertThresholds = Field(..., description="Alert thresholds")
    showNotifications: Optional[bool] = Field(True, description="Show browser notifications")
    playSound: Optional[bool] = Field(False, description="Play sound on alerts")
    flashWidget: Optional[bool] = Field(True, description="Flash widget on alerts")


class GaugeConfig(BaseModel):
    """Gauge widget specific configuration."""
    minValue: float = Field(..., description="Minimum gauge value")
    maxValue: float = Field(..., description="Maximum gauge value")
    startAngle: float = Field(..., ge=-360, le=360, description="Start angle in degrees")
    endAngle: float = Field(..., ge=-360, le=360, description="End angle in degrees") 
    arcWidth: float = Field(..., ge=1, le=50, description="Arc width in pixels")
    showValue: bool = Field(..., description="Show numeric value")
    showLabel: bool = Field(..., description="Show label text")
    showTicks: bool = Field(..., description="Show tick marks")
    tickInterval: float = Field(..., ge=1, le=100, description="Tick mark interval")
    unit: str = Field(..., max_length=10, description="Unit of measurement")
    
    @validator('maxValue')
    def validate_max_greater_than_min(cls, max_value, values):
        """Ensure max value is greater than min value."""
        if 'minValue' in values and max_value <= values['minValue']:
            raise ValueError("Max value must be greater than min value")
        return max_value


class GraphConfig(BaseModel):
    """Graph widget specific configuration."""
    timeRange: int = Field(..., ge=1, le=1440, description="Time range in minutes")
    maxDataPoints: int = Field(..., ge=10, le=1000, description="Maximum data points to display")
    showGrid: bool = Field(..., description="Show grid lines")
    showAxes: bool = Field(..., description="Show axis labels")
    lineWidth: float = Field(..., ge=1, le=10, description="Line width in pixels")
    fillArea: bool = Field(..., description="Fill area under line")
    smoothing: bool = Field(..., description="Apply line smoothing")
    yAxisMin: Optional[float] = Field(None, description="Y-axis minimum value")
    yAxisMax: Optional[float] = Field(None, description="Y-axis maximum value")
    
    @validator('yAxisMax')
    def validate_y_axis_range(cls, y_max, values):
        """Validate Y-axis range."""
        if y_max is not None and 'yAxisMin' in values and values['yAxisMin'] is not None:
            if y_max <= values['yAxisMin']:
                raise ValueError("Y-axis max must be greater than min")
        return y_max


class TextConfig(BaseModel):
    """Text widget specific configuration."""
    format: str = Field(..., min_length=1, max_length=100, description="Text format string")
    fontSize: float = Field(..., ge=8, le=72, description="Font size in pixels")
    fontWeight: str = Field(..., regex="^(normal|bold|light)$", description="Font weight")
    alignment: str = Field(..., regex="^(left|center|right)$", description="Text alignment")
    showIcon: bool = Field(..., description="Show icon")
    iconPosition: str = Field(..., regex="^(left|right|top|bottom)$", description="Icon position")
    iconSize: float = Field(..., ge=8, le=64, description="Icon size in pixels")


class MultiSensorConfig(BaseModel):
    """Multi-sensor widget specific configuration."""
    sensors: List[SensorType] = Field(..., min_items=1, max_items=10, description="Sensor types to display")
    layout: str = Field(..., regex="^(grid|list|radial)$", description="Layout style")
    showLabels: bool = Field(..., description="Show sensor labels")
    showValues: bool = Field(..., description="Show sensor values")
    compactMode: bool = Field(..., description="Use compact display mode")


class WidgetConfig(BaseModel):
    """Complete widget configuration."""
    id: str = Field(..., min_length=1, max_length=100, description="Unique widget identifier")
    type: WidgetType = Field(..., description="Widget type")
    title: str = Field(..., min_length=1, max_length=50, description="Widget title")
    sensorType: SensorType = Field(..., description="Primary sensor type")
    position: Position = Field(..., description="Widget position")
    size: Size = Field(..., description="Widget size")
    style: WidgetStyle = Field(..., description="Widget styling")
    alerts: WidgetAlerts = Field(..., description="Alert configuration")
    
    # Type-specific configurations (only one should be present based on type)
    gaugeConfig: Optional[GaugeConfig] = Field(None, description="Gauge-specific configuration")
    graphConfig: Optional[GraphConfig] = Field(None, description="Graph-specific configuration")
    textConfig: Optional[TextConfig] = Field(None, description="Text-specific configuration")
    multiSensorConfig: Optional[MultiSensorConfig] = Field(None, description="Multi-sensor-specific configuration")
    
    # Metadata
    createdAt: int = Field(..., description="Creation timestamp in milliseconds")
    updatedAt: int = Field(..., description="Last update timestamp in milliseconds")
    version: str = Field(..., regex=r'^\d+\.\d+\.\d+$', description="Configuration version")
    tags: List[str] = Field(..., max_items=10, description="Widget tags")
    description: Optional[str] = Field(None, max_length=200, description="Widget description")
    author: Optional[str] = Field(None, max_length=50, description="Widget author")
    
    # Runtime state (not serialized to JSON exports)
    isSelected: Optional[bool] = Field(None, description="Whether widget is selected")
    isResizing: Optional[bool] = Field(None, description="Whether widget is being resized")
    isDragging: Optional[bool] = Field(None, description="Whether widget is being dragged")
    zIndex: Optional[int] = Field(None, ge=0, le=1000, description="Z-index for layering")
    
    @validator('updatedAt')
    def validate_updated_after_created(cls, updated_at, values):
        """Ensure updated timestamp is after created timestamp."""
        if 'createdAt' in values and updated_at < values['createdAt']:
            raise ValueError("Updated timestamp must be after created timestamp")
        return updated_at
    
    @validator('type')
    def validate_type_specific_config(cls, widget_type, values):
        """Ensure type-specific configuration is present based on widget type."""
        # This will be validated in a root_validator since we need access to all fields
        return widget_type
    
    @validator('tags')
    def validate_tags(cls, tags):
        """Validate tag format."""
        for tag in tags:
            if len(tag) > 20 or not tag.strip():
                raise ValueError("Tags must be non-empty and max 20 characters")
        return tags
    
    class Config:
        """Pydantic configuration."""
        
        use_enum_values = True
        schema_extra = {
            "example": {
                "id": "widget-cpu-gauge-001",
                "type": "gauge",
                "title": "CPU Usage",
                "sensorType": "cpu",
                "position": {"x": 100, "y": 100},
                "size": {"width": 200, "height": 200},
                "style": {
                    "theme": "cyberpunk",
                    "colors": ["#00ffff", "#ff006e"],
                    "opacity": 1.0,
                    "borderRadius": 8,
                    "fontSize": 14,
                    "fontFamily": "Orbitron"
                },
                "alerts": {
                    "enabled": True,
                    "thresholds": {
                        "warning": 75,
                        "critical": 90
                    },
                    "showNotifications": True,
                    "playSound": False,
                    "flashWidget": True
                },
                "gaugeConfig": {
                    "minValue": 0,
                    "maxValue": 100,
                    "startAngle": -135,
                    "endAngle": 135,
                    "arcWidth": 10,
                    "showValue": True,
                    "showLabel": True,
                    "showTicks": True,
                    "tickInterval": 20,
                    "unit": "%"
                },
                "createdAt": 1640995200000,
                "updatedAt": 1640995200000,
                "version": "1.0.0",
                "tags": ["cpu", "performance", "monitoring"],
                "description": "Real-time CPU usage monitoring gauge",
                "author": "SenseCanvas"
            }
        }


class GridSize(BaseModel):
    """Dashboard grid configuration."""
    cols: int = Field(..., ge=1, le=20, description="Number of grid columns")
    rows: int = Field(..., ge=1, le=20, description="Number of grid rows")
    cellSize: int = Field(..., ge=50, le=200, description="Grid cell size in pixels")


class DashboardLayout(BaseModel):
    """Complete dashboard layout configuration."""
    id: str = Field(..., min_length=1, max_length=100, description="Unique layout identifier")
    name: str = Field(..., min_length=1, max_length=50, description="Layout name")
    description: str = Field(..., max_length=200, description="Layout description")
    widgets: List[WidgetConfig] = Field(..., max_items=50, description="Widget configurations")
    gridSize: GridSize = Field(..., description="Grid configuration")
    theme: ThemeType = Field(..., description="Default theme")
    backgroundImage: Optional[str] = Field(None, description="Background image URL")
    backgroundColor: Optional[str] = Field(None, regex="^#[0-9A-Fa-f]{6}$", description="Background color")
    
    # Metadata
    createdAt: int = Field(..., description="Creation timestamp in milliseconds")
    updatedAt: int = Field(..., description="Last update timestamp in milliseconds")
    version: str = Field(..., regex=r'^\d+\.\d+\.\d+$', description="Layout version")
    author: Optional[str] = Field(None, max_length=50, description="Layout author")
    tags: List[str] = Field(..., max_items=10, description="Layout tags")
    
    # Export/Import
    isPublic: bool = Field(..., description="Whether layout is public")
    shareKey: Optional[str] = Field(None, description="Sharing key for public layouts")
    
    @validator('widgets')
    def validate_widget_ids_unique(cls, widgets):
        """Ensure all widget IDs are unique within the layout."""
        widget_ids = [w.id for w in widgets]
        if len(widget_ids) != len(set(widget_ids)):
            raise ValueError("Widget IDs must be unique within a layout")
        return widgets
    
    class Config:
        """Pydantic configuration."""
        
        use_enum_values = True
        schema_extra = {
            "example": {
                "id": "layout-gaming-001",
                "name": "Gaming Dashboard",
                "description": "Optimized layout for gaming performance monitoring",
                "widgets": [],
                "gridSize": {
                    "cols": 12,
                    "rows": 8,
                    "cellSize": 80
                },
                "theme": "gaming",
                "backgroundColor": "#0a0a0a",
                "createdAt": 1640995200000,
                "updatedAt": 1640995200000,
                "version": "1.0.0",
                "author": "SenseCanvas",
                "tags": ["gaming", "performance", "rgb"],
                "isPublic": True,
                "shareKey": "gaming-dash-001"
            }
        }


class WidgetLibraryItem(BaseModel):
    """Widget library/preset item."""
    id: str = Field(..., min_length=1, max_length=100, description="Library item identifier")
    name: str = Field(..., min_length=1, max_length=50, description="Widget name")
    description: str = Field(..., min_length=1, max_length=200, description="Widget description")
    category: str = Field(..., min_length=1, max_length=30, description="Widget category")
    tags: List[str] = Field(..., max_items=10, description="Widget tags")
    thumbnail: str = Field(..., description="Thumbnail image URL or data URI")
    config: WidgetConfig = Field(..., description="Widget configuration")
    popularity: float = Field(..., ge=0, le=100, description="Popularity score")
    author: str = Field(..., max_length=50, description="Widget author")
    createdAt: int = Field(..., description="Creation timestamp")
    downloads: int = Field(..., ge=0, description="Download count")


class AiGenerationPrompt(BaseModel):
    """AI widget generation prompt and result."""
    id: str = Field(..., min_length=1, max_length=100, description="Prompt identifier")
    prompt: str = Field(..., min_length=10, max_length=500, description="User prompt")
    context: Dict[str, Any] = Field(..., description="Generation context")
    generatedConfig: Dict[str, Any] = Field(..., description="Generated widget configuration")
    rating: Optional[int] = Field(None, ge=1, le=5, description="User rating")
    createdAt: int = Field(..., description="Creation timestamp")


class WidgetExportData(BaseModel):
    """Widget configuration export data."""
    version: str = Field(..., regex=r'^\d+\.\d+\.\d+$', description="Export format version")
    exportedAt: int = Field(..., description="Export timestamp")
    exportedBy: Optional[str] = Field(None, max_length=50, description="Exporter name")
    widgets: List[WidgetConfig] = Field(..., min_items=1, max_items=50, description="Exported widgets")
    layout: Optional[DashboardLayout] = Field(None, description="Optional layout information")


class ConfigImportOptions(BaseModel):
    """Configuration import options."""
    preserveIds: Optional[bool] = Field(False, description="Preserve original widget IDs")
    replaceExisting: Optional[bool] = Field(False, description="Replace existing widgets with same IDs")
    validateCompatibility: Optional[bool] = Field(True, description="Validate version compatibility")